File: zig/plugins/aws.zig
//! AWS Secrets Manager and SSM Parameter Store - one sekreto plugin,
//! two kinds: `awssecrets` and `awsparams`.
//!
//! Requests are SigV4-signed in-tree (sigv4.zig, beside this file): the
//! HMAC-SHA256 that needs lives here and not in the core, so a chain
//! that never names AWS never links a hash function.

const std = @import("std");

const sekreto = @import("sekreto");
const httpjson = @import("httpjson.zig");
const sigv4 = @import("sigv4.zig");

const Allocator = std.mem.Allocator;
const Answer = sekreto.Answer;
const Found = sekreto.Found;
const Provider = sekreto.Provider;
const ProviderSpec = sekreto.ProviderSpec;

/// Region and credentials, from config first and the standard AWS_*
/// environment variables second. Missing either is an error: an AWS store
/// with no credentials could not answer.
const AwsAuth = struct {
    region: []const u8,
    keyid: []const u8,
    secret: []const u8,
    session: []const u8,
};

/// The common half of both AWS providers.
const Aws = struct {
    config: sekreto.Config,
    region: []const u8,
    keyid: []const u8,
    secret: []const u8,
    session: []const u8,
    addr: []const u8,
    prefix: []const u8,

    fn auth(self: Aws, alloc: Allocator) Allocator.Error!Answer(AwsAuth) {
        const region = httpjson.firstof(self.config, self.region, &.{ "AWS_REGION", "AWS_DEFAULT_REGION" });
        const keyid = httpjson.firstof(self.config, self.keyid, &.{"AWS_ACCESS_KEY_ID"});
        const secret = httpjson.firstof(self.config, self.secret, &.{"AWS_SECRET_ACCESS_KEY"});
        const session = httpjson.firstof(self.config, self.session, &.{"AWS_SESSION_TOKEN"});

        if (0 == region.len) {
            return .{ .err = try sekreto.fail(
                alloc,
                "sekreto: aws: no region (set region or AWS_REGION)",
                .{},
            ) };
        }

        if (0 == keyid.len or 0 == secret.len) {
            return .{ .err = try sekreto.fail(
                alloc,
                "sekreto: aws: no credentials (set keyid/secret or AWS_ACCESS_KEY_ID/AWS_SECRET_ACCESS_KEY)",
                .{},
            ) };
        }

        return .{ .ok = .{ .region = region, .keyid = keyid, .secret = secret, .session = session } };
    }

    /// One signed call to an AWS JSON-1.1 API.
    fn call(
        self: Aws,
        alloc: Allocator,
        service: []const u8,
        target: []const u8,
        payload: []const u8,
    ) Allocator.Error!httpjson.Result {
        const credentials = switch (try self.auth(alloc)) {
            .err => |message| return .{ .err = message },
            .ok => |got| got,
        };

        // The China partition lives under its own suffix; every other
        // commercial region is plain amazonaws.com.
        const suffix: []const u8 = if (std.mem.startsWith(u8, credentials.region, "cn-"))
            ".amazonaws.com.cn"
        else
            ".amazonaws.com";

        const addr = if (0 != self.addr.len)
            self.addr
        else
            try std.fmt.allocPrint(alloc, "https://{s}.{s}{s}", .{ service, credentials.region, suffix });

        switch (try sekreto.checkaddr(alloc, addr)) {
            .err => |message| return .{ .err = message },
            .ok => {},
        }

        const url = try std.fmt.allocPrint(alloc, "{s}/", .{httpjson.trimslash(addr)});

        const unsigned = [_]sigv4.Pair{
            .{ .name = "content-type", .value = "application/x-amz-json-1.1" },
            .{ .name = "x-amz-target", .value = target },
        };

        const signed = try sigv4.sign(alloc, .{
            .method = "POST",
            .url = url,
            .headers = &unsigned,
            .body = payload,
            .service = service,
            .region = credentials.region,
            .keyid = credentials.keyid,
            .secret = credentials.secret,
            .session = credentials.session,
            .datetime = try awsnow(alloc, self.config.io),
        });

        var headers: std.ArrayList(httpjson.Header) = .empty;
        for (unsigned) |header| {
            try headers.append(alloc, .{ .name = header.name, .value = header.value });
        }
        for (signed) |header| {
            try headers.append(alloc, .{ .name = header.name, .value = header.value });
        }

        return httpjson.fetchjson(alloc, self.config.io, .POST, url, headers.items, payload);
    }
};

/// The `YYYYMMDDTHHMMSSZ` timestamp SigV4 wants, for now.
fn awsnow(alloc: Allocator, io: std.Io) Allocator.Error![]const u8 {
    const seconds: u64 = @intCast(@divTrunc(httpjson.nowms(io), 1000));

    const day = std.time.epoch.EpochSeconds{ .secs = seconds };
    const yearday = day.getEpochDay().calculateYearDay();
    const monthday = yearday.calculateMonthDay();
    const clock = day.getDaySeconds();

    return std.fmt.allocPrint(alloc, "{d:0>4}{d:0>2}{d:0>2}T{d:0>2}{d:0>2}{d:0>2}Z", .{
        yearday.year,
        monthday.month.numeric(),
        monthday.day_index + 1,
        clock.getHoursIntoDay(),
        clock.getMinutesIntoHour(),
        clock.getSecondsIntoMinute(),
    });
}

/// Does this AWS error body name one of the not-found types? Those are a
/// miss; every other failure is a store that could not answer.
fn awsmiss(body: ?std.json.Value, want: []const u8) bool {
    const errtype = httpjson.jstr(httpjson.jget(body, "__type")) orelse return false;
    return null != std.mem.indexOf(u8, errtype, want);
}

/// AWS Secrets Manager.
///
/// `api.token` reads the secret named `api` (the vaultref path, so
/// `db.pass.main` reads `db/pass`) and takes the `token` field of its JSON
/// SecretString - the AWS idiom of one JSON map per secret. A SecretString
/// that is not JSON is the value itself, under the conventional field
/// `value`. Requests are SigV4-signed in-tree; see sigv4.zig.
pub const AwsSecretsProvider = struct {
    aws: Aws,

    pub fn lookup(self: *AwsSecretsProvider, alloc: Allocator, name: []const u8) Allocator.Error!Found {
        const ref = switch (try sekreto.vaultref(alloc, name)) {
            .err => |message| return .{ .err = message },
            .ok => |got| got,
        };

        const payload = try httpjson.jsonobject(alloc, &.{.{ .key = "SecretId", .value = ref.path }});

        const res = switch (try self.aws.call(alloc, "secretsmanager", "secretsmanager.GetSecretValue", payload)) {
            .err => |message| return .{ .err = message },
            .ok => |got| got,
        };

        if (400 == res.status and awsmiss(res.body, "ResourceNotFoundException")) {
            return .{ .ok = null };
        }

        if (200 != res.status) {
            return .{ .err = try sekreto.fail(
                alloc,
                "sekreto: aws secretsmanager error: {d}",
                .{res.status},
            ) };
        }

        const text = httpjson.jstr(httpjson.jget(res.body, "SecretString")) orelse {
            // A binary secret has no fields to address; only the
            // conventional `value` field can mean "the bytes themselves".
            const binary = httpjson.jstr(httpjson.jget(res.body, "SecretBinary")) orelse return .{ .ok = null };
            if (!std.mem.eql(u8, "value", ref.field)) {
                return .{ .ok = null };
            }
            // A payload that will not decode was reported as null - a
            // MISS - so the chain carried on to a weaker store. A store
            // that answered incoherently could not answer.
            return .{ .ok = try sekreto.unbase64(alloc, binary) orelse
                return .{ .err = try sekreto.fail(alloc, "sekreto: aws secretsmanager: undecodable secret", .{}) } };
        };

        const parsed = std.json.parseFromSlice(std.json.Value, alloc, text, .{}) catch {
            // A plain-string secret is the whole value; it has no named
            // fields.
            return .{ .ok = if (std.mem.eql(u8, "value", ref.field)) text else null };
        };

        if (.object == parsed.value) {
            return .{ .ok = httpjson.jstr(httpjson.jget(parsed.value, ref.field)) };
        }

        return .{ .ok = if (std.mem.eql(u8, "value", ref.field)) text else null };
    }

    pub fn describe(self: *AwsSecretsProvider, alloc: Allocator) Allocator.Error![]const u8 {
        // sekreto.Config only, never the environment: describe() feeds the spec's
        // sources group, which must answer the same everywhere.
        return std.fmt.allocPrint(alloc, "awssecrets:{s}", .{self.aws.region});
    }

    pub fn deinit(_: *AwsSecretsProvider, _: Allocator) void {}
};

/// AWS SSM Parameter Store.
///
/// `db.pass.main` reads the parameter `/db/pass/main` (under an optional
/// prefix path), decrypted. Parameter Store carries flat strings, so there
/// is no field indirection.
pub const AwsParamsProvider = struct {
    aws: Aws,

    pub fn lookup(self: *AwsParamsProvider, alloc: Allocator, name: []const u8) Allocator.Error!Found {
        const param = switch (try sekreto.awsparam(alloc, name, self.aws.prefix)) {
            .err => |message| return .{ .err = message },
            .ok => |got| got,
        };

        const payload = try std.fmt.allocPrint(
            alloc,
            "{{\"Name\":{s},\"WithDecryption\":true}}",
            .{try httpjson.jsonstring(alloc, param)},
        );

        const res = switch (try self.aws.call(alloc, "ssm", "AmazonSSM.GetParameter", payload)) {
            .err => |message| return .{ .err = message },
            .ok => |got| got,
        };

        if (400 == res.status and awsmiss(res.body, "ParameterNotFound")) {
            return .{ .ok = null };
        }

        if (200 != res.status) {
            return .{ .err = try sekreto.fail(alloc, "sekreto: aws ssm error: {d}", .{res.status}) };
        }

        return .{ .ok = httpjson.jstr(httpjson.jget(httpjson.jget(res.body, "Parameter"), "Value")) };
    }

    pub fn describe(self: *AwsParamsProvider, alloc: Allocator) Allocator.Error![]const u8 {
        return std.fmt.allocPrint(alloc, "awsparams:{s}{s}", .{ self.aws.region, self.aws.prefix });
    }

    pub fn deinit(_: *AwsParamsProvider, _: Allocator) void {}
};

fn awsof(config: sekreto.Config, spec: ProviderSpec) Aws {
    return .{
        .config = config,
        .region = spec.region,
        .keyid = spec.keyid,
        .secret = spec.secret,
        .session = spec.session,
        .addr = spec.addr,
        .prefix = spec.prefix,
    };
}

fn makesecrets(alloc: Allocator, config: sekreto.Config, spec: ProviderSpec) Allocator.Error!Answer(Provider) {
    return .{ .ok = try sekreto.provide(alloc, AwsSecretsProvider, .{ .aws = awsof(config, spec) }) };
}

fn makeparams(alloc: Allocator, config: sekreto.Config, spec: ProviderSpec) Allocator.Error!Answer(Provider) {
    return .{ .ok = try sekreto.provide(alloc, AwsParamsProvider, .{ .aws = awsof(config, spec) }) };
}

/// The `awssecrets` kind.
pub const awssecrets: sekreto.Definition = sekreto.providerplugin("awssecrets", makesecrets);

/// The `awsparams` kind.
pub const awsparams: sekreto.Definition = sekreto.providerplugin("awsparams", makeparams);

File: php/plugins/aws.php
<?php

/**
 * AWS Secrets Manager and SSM Parameter Store, as voxgig/plugin
 * definitions - TWO kinds from one file, because they share request signing
 * and the credential chain.
 *
 * A PLUGIN, not a built-in, and the sharpest instance of why: it opens a
 * socket AND it signs, so `sigv4.php` - the only hashing in this port - is
 * required from here and from nowhere in the core.
 */

declare(strict_types=1);

namespace Voxgig\Sekreto\Plugins;

require_once __DIR__ . '/../src/Providers.php';
require_once __DIR__ . '/../src/Addr.php';
require_once __DIR__ . '/httpjson.php';
require_once __DIR__ . '/sigv4.php';

use Voxgig\Sekreto\Name;
use Voxgig\Sekreto\Provider;
use Voxgig\Sekreto\SekretoError;

use function Voxgig\Sekreto\checkaddr;
use function Voxgig\Sekreto\providerplugin;

/** The `YYYYMMDDTHHMMSSZ` timestamp SigV4 wants, for now. */
function awsnow(): string
{
    return gmdate('Ymd\THis\Z');
}

/**
 * Region and credentials, from config first and the standard AWS_*
 * environment variables second - those are AWS's own convention, and a
 * pod or CI job that has them set should just work. Missing either is an
 * error: an AWS store with no credentials could not answer.
 *
 * @param array<string, mixed> $opts
 * @return array{region: string, keyid: string, secret: string, session: ?string}
 */
function awsauth(array $opts): array
{
    $region = ($opts['region'] ?? null)
        ?: (getenv('AWS_REGION') ?: (getenv('AWS_DEFAULT_REGION') ?: ''));
    $keyid = ($opts['keyid'] ?? null) ?: (getenv('AWS_ACCESS_KEY_ID') ?: '');
    $secret = ($opts['secret'] ?? null) ?: (getenv('AWS_SECRET_ACCESS_KEY') ?: '');
    $session = ($opts['session'] ?? null) ?: (getenv('AWS_SESSION_TOKEN') ?: null);

    if ('' === $region) {
        throw new SekretoError('sekreto: aws: no region (set region or AWS_REGION)');
    }
    if ('' === $keyid || '' === $secret) {
        throw new SekretoError(
            'sekreto: aws: no credentials'
            . ' (set keyid/secret or AWS_ACCESS_KEY_ID/AWS_SECRET_ACCESS_KEY)'
        );
    }

    return ['region' => $region, 'keyid' => $keyid, 'secret' => $secret, 'session' => $session];
}

/**
 * One signed call to an AWS JSON-1.1 API.
 *
 * @param array<string, mixed> $opts
 * @param array<string, mixed> $payload
 * @return array{0: int, 1: mixed}
 */
function awscall(array $opts, string $service, string $target, array $payload): array
{
    $auth = awsauth($opts);
    // The China partition lives under its own suffix; every other
    // commercial region is plain amazonaws.com.
    $suffix = str_starts_with($auth['region'], 'cn-') ? '.amazonaws.com.cn' : '.amazonaws.com';
    $addr = ($opts['addr'] ?? null)
        ?: 'https://' . $service . '.' . $auth['region'] . $suffix;
    checkaddr($addr);

    $url = preg_replace('#/$#', '', $addr) . '/';
    // Unescaped slashes, so the body signed here is byte-identical to the
    // canonical port's JSON.stringify output.
    $body = json_encode($payload, JSON_UNESCAPED_SLASHES);
    $headers = [
        'content-type' => 'application/x-amz-json-1.1',
        'x-amz-target' => $target,
    ];

    $signed = sigv4([
        'method' => 'POST',
        'url' => $url,
        'headers' => $headers,
        'body' => $body,
        'service' => $service,
        'region' => $auth['region'],
        'keyid' => $auth['keyid'],
        'secret' => $auth['secret'],
        'session' => $auth['session'],
        'datetime' => awsnow(),
    ]);

    $lines = [];
    foreach (array_merge($headers, $signed) as $key => $value) {
        $lines[] = $key . ': ' . $value;
    }

    return fetchjson('POST', $url, $lines, $body);
}

/**
 * Does this AWS error body name one of the not-found types? Those are a
 * miss; every other failure is a store that could not answer.
 *
 * @param array<int, string> $types
 */
function awsmiss(mixed $body, array $types): bool
{
    $errtype = is_array($body) && is_string($body['__type'] ?? null) ? $body['__type'] : '';

    foreach ($types as $name) {
        if (str_contains($errtype, $name)) {
            return true;
        }
    }

    return false;
}

/**
 * AWS Secrets Manager.
 *
 * `api.token` reads the secret named `api` (the vaultref path, so
 * `db.pass.main` reads `db/pass`) and takes the `token` field of its JSON
 * SecretString - the AWS idiom of one JSON map per secret. A SecretString
 * that is not JSON is the value itself, under the conventional field
 * `value`. Requests are SigV4-signed in-tree; see sigv4.php.
 */
class AwsSecretsProvider implements Provider
{
    /** @param array<string, mixed> $opts */
    public function __construct(private array $opts = [])
    {
    }

    public function lookup(string $name): ?string
    {
        $ref = Name::vaultref($name);

        [$status, $body] = awscall(
            $this->opts,
            'secretsmanager',
            'secretsmanager.GetSecretValue',
            ['SecretId' => $ref['path']]
        );

        if (400 === $status && awsmiss($body, ['ResourceNotFoundException'])) {
            return null;
        }

        if (200 !== $status) {
            throw new SekretoError('sekreto: aws secretsmanager error: ' . $status);
        }

        $text = $body['SecretString'] ?? null;

        if (!is_string($text)) {
            // A binary secret has no fields to address; only the conventional
            // `value` field can mean "the bytes themselves".
            $bin = $body['SecretBinary'] ?? null;
            if (is_string($bin) && 'value' === $ref['field']) {
                // strict: base64_decode SKIPS characters outside the
                // alphabet by default, so a corrupted payload decoded to
                // plausible-looking bytes that were then returned as the
                // secret. A store that answered incoherently is an error.
                $decoded = base64_decode($bin, true);
                if (false === $decoded) {
                    throw new SekretoError('sekreto: aws secretsmanager: undecodable secret');
                }

                return $decoded;
            }
            return null;
        }

        // Decoded without assoc, so a JSON object is distinguishable from a
        // JSON array: only an object carries named fields.
        $parsed = json_decode($text);

        if ($parsed instanceof \stdClass) {
            $value = $parsed->{$ref['field']} ?? null;
            return null === $value ? null : (string) $value;
        }

        // A plain-string secret is the whole value; it has no named fields.
        return 'value' === $ref['field'] ? $text : null;
    }

    // Config only, never the environment: describe() feeds the spec's
    // sources group, which must answer the same everywhere.
    public function describe(): string
    {
        return 'awssecrets:' . ($this->opts['region'] ?? '');
    }
}

/**
 * AWS SSM Parameter Store.
 *
 * `db.pass.main` reads the parameter `/db/pass/main` (under an optional
 * prefix path), decrypted. Parameter Store carries flat strings, so there
 * is no field indirection.
 */
class AwsParamsProvider implements Provider
{
    /** @param array<string, mixed> $opts */
    public function __construct(private array $opts = [])
    {
    }

    public function lookup(string $name): ?string
    {
        [$status, $body] = awscall($this->opts, 'ssm', 'AmazonSSM.GetParameter', [
            'Name' => Name::awsparam($name, $this->opts['prefix'] ?? null),
            'WithDecryption' => true,
        ]);

        if (400 === $status && awsmiss($body, ['ParameterNotFound'])) {
            return null;
        }

        if (200 !== $status) {
            throw new SekretoError('sekreto: aws ssm error: ' . $status);
        }

        $value = $body['Parameter']['Value'] ?? null;

        return null === $value ? null : (string) $value;
    }

    public function describe(): string
    {
        return 'awsparams:' . ($this->opts['region'] ?? '') . ($this->opts['prefix'] ?? '');
    }
}

/**
 * The `awssecrets` provider kind, as a voxgig/plugin definition. Pass it to
 * `Sekreto` in the `plugins` option; nothing else loads it.
 *
 * @return array<string, mixed>
 */
function awssecrets(): array
{
    return providerplugin('awssecrets', fn(array $spec) => new AwsSecretsProvider($spec));
}

/**
 * The `awsparams` provider kind, as a voxgig/plugin definition. Pass it to
 * `Sekreto` in the `plugins` option; nothing else loads it.
 *
 * @return array<string, mixed>
 */
function awsparams(): array
{
    return providerplugin('awsparams', fn(array $spec) => new AwsParamsProvider($spec));
}
